File: C++/18. 4Sum.cpp
vector<vector<int>> fourSum(vector<int>& nums, int target) 
{
    sort(nums.begin(), nums.end()); 
    return kSum(nums, target, 0, 4); 
}

vector<vector<int>> kSum(vector<int>& nums, int target, int start, int k)
{
    vector<vector<int>> res ;
    if(start == nums.size() or static_cast<long>(nums[start]) * k > target or target > long(nums.back()*4))
        return res ;

    if(k == 2)
        return twoSum(nums, target, start); 

    for(int i = start ; i < nums.size(); i++)
    {
        if(i > start and nums[i] == nums[i-1]) // handle Duplicates 
            continue ; 

        for(vector<int>& subRes : kSum(nums, target-nums[i], i+1, k-1))
        {
            subRes.push_back(nums[i]);
            res.push_back(subRes);  
        }
    }

    return res ; 
}

vector<vector<int>> twoSum(vector<int>& nums, int target, int start)
{
    vector<vector<int>> res ;
    int i = start, j = nums.size()-1, sum ; 

    while( i < j )
    {
        sum = nums[i] + nums[j] ; 

        if(sum < target or (i > start and nums[i] == nums[i-1]))
            i++ ; 
        else if(sum > target or (j < nums.size()-1 and nums[j] == nums[j+1]))
            j-- ; 
        else
            res.push_back({nums[i++], nums[j--]}); 
    }

    return res ; 
}

File: Rust/18. 4Sum.rs
// Time Limit Exceed
use std::collections::HashSet;

impl Solution {
    pub fn four_sum(mut nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        nums.sort();
        let mut set: HashSet<Vec<i32>> = HashSet::new();

        if nums.len() < 4 {
            return Vec::new();
        }

        for i in 0..nums.len() - 3 {
            for j in i + 1..nums.len() - 2 {
                for k in j + 1..nums.len() - 1 {
                    for l in k + 1..nums.len() {
                        if nums[i] + nums[j] + nums[k] + nums[l] == target {
                            set.insert(vec![nums[i], nums[j], nums[k], nums[l]]);
                        }
                    }
                }
            }
        }
        set.into_iter().collect()
    }
}

impl Solution {
    pub fn four_sum(mut nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        if nums.len() < 4 {
            return Vec::new();
        }

        let mut set: HashSet<Vec<i32>> = HashSet::new();
        let mut left: usize;
        let mut right: usize;

        nums.sort();

        for i in 0..nums.len() - 3 {
            // skip the repeats
            if i > 0 && nums[i] == nums[i - 1] {
                continue;
            }

            for j in i + 1..nums.len() - 2 {
                left = j + 1;
                right = nums.len() - 1;

                while left < nums.len() && left < right {
                    match (nums[i] + nums[j] + nums[left] + nums[right]) as i64 {
                        sum if sum == target as i64 => {
                            set.insert(vec![nums[i], nums[j], nums[left], nums[right]]);

                            while left < nums.len() - 1 && nums[left] == nums[left + 1] {
                                left += 1;
                            }

                            while right > 0 && nums[right] == nums[right - 1] {
                                right -= 1;
                            }

                            right -= 1;
                            left += 1;
                        }
                        sum if sum > target as i64 => {
                            right -= 1;
                        }
                        _ => {
                            left += 1;
                        }
                    }
                }
            }
        }
        set.into_iter().collect()
    }
}
